### PR TITLE
add annotation option for serialization

### PR DESCRIPTION
**Pull request by elibixby**
# TL;DR
Allow users to specify TypeTransformers via Annotation

This is a followup to conversation in https://github.com/flyteorg/flytekit/pull/1565 

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [ ] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
https://github.com/flyteorg/flyte/issues/<number>

### CODE REVIEW

**Comment by elibixby** [comment]
@wild-endeavor @pingsutw 

Based on our usage at Cradle I think this is the highest priority improvement to come out of https://github.com/flyteorg/flytekit/pull/1565 (which is a little overstuffed :) )

A pattern we find ourselves reaching pretty frequently for is:

```
class SerializationWrapper:

    def __class_getitem__(cls, wrapped_type):
        TypeEngine.register(Transformer(wrapped_type), wrapped_type)
        return wrapped_type
```

However this is limited if we want to specify alternate serialization for types, as `TypeEngine` enforces a 1:1 mapping from python types to `TypeTransformer`. 

This small change unlocks that requirement, by storing the type serialization information in the annotation of the type.

**Comment by wild-endeavor** [comment]
So the idea is that we have

```python
class MyType(object):
    ...

class MyTypeTransformer(TypeTransformer[MyType]):
    ...

class SecondaryMyTypeTransformer(TypeTransformer[MyType]):
    ...

MM = Annotated[MyType, SecondaryMyTypeTransformer]
m1 = MyType()
m2 = MM()
```
then `m1` would get the first transformer and `m2` would get the second transformer right?  I think in spirit I'm mostly okay with this because it's like a manual bypass for the type engine.  The user is overriding the type engine and declaring what transformer should be used.

I don't think it should be saved though, or added to guess_python_types.  Honestly the guessing bit has led to a lot of frustration on the user side especially as we've added more complex types like `StructuredDataset`.  We eventually gave up and just added the ability to specify a type to the `LiteralsResolver` object.

Would like to get @eapolinario's thoughts on this though.

**Comment by pingsutw** [comment]
qq: when do you use two transformer for certain type? why can't just override default flytekit transformer?

**Comment by elibixby** [comment]
> qq: when do you use two transformer for certain type? why can't just override default flytekit transformer?

I can think there are two main cases:

1. You have a generic type which has many possible sources 
2. A base type where you don't want to override the implementation but it comes from another source

In general, these boil down to the same thing: explicitly separating the *source* of the data (as a generic, or type factory), and the *type* of the data when it is decoded. I can see this being used in `FlyteFile`, `StructuredDataset` and essentially any other non-primitive type, to create very useful and intuitive behavior. 

For example:

```
class StructuredDataset:

    def __class_getitem__(self, decoded_as: Type[T], kwtypes: OrderedDict[str, type]) -> Type[T]:
         return Annotated[decoded_as, StructuredDatasetTypeTransformer(decoded_as=decoded_as, schema=kwtypes)]
 
 
schema = kwtypes(foo=int, bar=str)
 
@task
def my_task_dataframe(dataset: StructuredDataset[pd.DataFrame, schema]):
    assert isinstance(dataset, pd.DataFrame)
    
@task
def my_task_dataset(dataset: StructuredDataset[torch.data.Dataset, schema]):
   ...
   
@task
def my_task_file(dataset: StructuredDataset[TextIO, schema]):
   ...
   
@task
def my_task_filepath(dataset: StructuredDataset[str, schema]):
   ...
 
```

In particular, the last two are not possible under the current system, because not every `TextIO` or `str` would be applicable, (you'd only want files or file-paths that represent parquet)

At cradle we want to be able to chain together workflows while keeping provenance information. We've decided the easiest way to do this is to have an `OutputRef` type which uses `FlyteRemote` to fetch the output of a specified execution from the Flyte server. The ideal syntax here is `OutputRef[MyType]`, which inside the function is treated as `MyType`. But this doesn't work because by definition `MyType` will already have a registered type transformer (it is the output of another workflow after all).

We also would like to be able to pass references to tasks/launch plans/workflows into workflow functions for incorporation into the dag, as part of a `dynamic` call. The ideal approach here would be similar `RemoteRef[Callable[...]]`, but registering generics dynamically is currently not possible.

File: flytekit/core/type_engine.py [hunk 1]
```diff
@@ -929,6 +936,13 @@ def guess_python_type(cls, flyte_type: LiteralType) -> type:
         """
         Transforms a flyte-specific ``LiteralType`` to a regular python value.
         """
+        # Try extra transformers first so overridden registered types, aren't improperly guessed
+        for transformer in cls._EXTRA_TRANSFORMERS:
+            try:
+                return transformer.guess_python_type(flyte_type)
+            except ValueError:
+                logger.debug(f"Skipping transformer {transformer.name} for {flyte_type}")
```
[the remark this answers is not in the document]
Comment (elibixby) [reply flytekit/core/type_engine.py:944]: Great, I wasn't sure if `guess_python_type` was required behavior, but indeed if it is deprecated going forward it makes the annotations stateless, which is desirable I think.

[the remark this answers is not in the document]
Comment (elibixby) [reply flytekit/core/type_engine.py]: Done

**Review by wild-endeavor — APPROVED** [review]

**Review by pingsutw — DISMISSED** [review]

File: tests/flytekit/unit/core/test_type_engine.py [hunk 2]
```diff
@@ -170,6 +172,51 @@ class Foo(object):
...
+        def to_literal(
+            self, ctx: FlyteContext, python_val: T, python_type: typing.Type[T], expected: LiteralType
+        ) -> Literal:
+            return Literal(scalar=Scalar(primitive=Primitive(string_value=json.dumps(python_val))))
+
+    class JSONSerialized:
+        def __class_getitem__(cls, item: Type[T]):
+            return Annotated[item, JsonTypeTransformer(name=f"json[{item}]", t=item)]
+
+    MyJsonDict = JSONSerialized[dict[str, int]]
```
Comment (wild-endeavor) [inline tests/flytekit/unit/core/test_type_engine.py:196]:
```suggestion
    MyJsonDict = JSONSerialized[typing.Dict[str, int]]
```

**Review by wild-endeavor — APPROVED** [review]

**Review by pingsutw — APPROVED** [review]

**Review by pingsutw — APPROVED** [review]

**Pull request merged**